tests/src/test/java/alluxio/collections/IndexedSetIntegrationTest.java

```diff
@@ -0,0 +1,259 @@
+      new IndexedSet.FieldIndex<TestInfo>() {
+        @Override
+        public Object getFieldValue(TestInfo o) {
+          return o.getSize();
+        }
+      };
+
+  @Before
+  public void before() throws Exception {
+    mIndexedSet = new IndexedSet<>(mIdIndex, mSizeIndex);
+    mThreadPool = Executors.newCachedThreadPool();
+    mStopThreads = new AtomicBoolean(false);
+  }
+
+  @After
+  public void after() {
+    mThreadPool.shutdownNow();
+  }
+
+  /**
+   * Verifies the {@link #mIndexedSet} for internal consistency.
+   */
+  private void verifySet() {
+    Iterator<TestInfo> it = mIndexedSet.iterator();
+    Set<Long> ids = new HashSet<>();
+    Set<Integer> sizes = new HashSet<>();
+
+    // Verify the size.
+    int expectedCount = 0;
+    while (it.hasNext()) {
+      TestInfo info = it.next();
+      ids.add(info.getId());
+      sizes.add(info.getSize());
+      expectedCount++;
+    }
+    Assert.assertEquals(expectedCount, mIndexedSet.size());
+
+    // Verify the size according to the id index.
+    int count = 0;
+    for (Long id : ids) {
+      Set<TestInfo> elements = mIndexedSet.getByField(mIdIndex, id);
+      count += elements.size();
+    }
+    Assert.assertEquals(expectedCount, count);
+
+    // Verify the size according to the size index.
+    count = 0;
+    for (Integer size : sizes) {
+      Set<TestInfo> elements = mIndexedSet.getByField(mSizeIndex, size);
+      count += elements.size();
+    }
+    Assert.assertEquals(expectedCount, count);
+  }
+
+  @Test
+  public void basicConcurrentTest() throws Exception {
+    List<Future<?>> futures = new ArrayList<>();
+    List<ConcurrentTask> addTasks = new ArrayList<>();
+    List<ConcurrentTask> removeTasks = new ArrayList<>();
+
+    addTasks.add(new ConcurrentAdd());
+    addTasks.add(new ConcurrentAdd());
+    addTasks.add(new ConcurrentAdd());
+    addTasks.add(new ConcurrentAdd());
+    removeTasks.add(new ConcurrentRemove());
+    removeTasks.add(new ConcurrentRemove());
+    removeTasks.add(new ConcurrentRemoveByField());
+    removeTasks.add(new ConcurrentRemoveByField());
+    for (ConcurrentTask task : addTasks) {
+      futures.add(mThreadPool.submit(task));
+    }
+    for (ConcurrentTask task : removeTasks) {
+      futures.add(mThreadPool.submit(task));
+    }
+
+    CommonUtils.sleepMs(TEST_CASE_DURATION_MS);
+    mStopThreads.set(true);
+    for (Future<?> future : futures) {
+      future.get();
+    }
+
+    // Calculate how many elements have been added or removed.
+    long added = 0;
+    for (ConcurrentTask task : addTasks) {
+      added += task.getCount();
+    }
+    long removed = 0;
+    for (ConcurrentTask task : removeTasks) {
+      removed += task.getCount();
+    }
+
+    Assert.assertEquals(mIndexedSet.size(), added - removed);
+    verifySet();
+  }
+
+  @Test
+  public void concurrentTest() throws Exception {
+    List<Future<?>> futures = new ArrayList<>();
+
+    futures.add(mThreadPool.submit(new ConcurrentAdd()));
```

Comment:
it would be valuable to randomize the number of different tasks across different test executions